File: go/0150M-逆波兰表达式求值.go
/*
 * @Author: SukiEva
 * @Date: 2021-10-23 16:23:29
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */

package main

import "strconv"

func evalRPN(tokens []string) int {
	stack := make([]int, 0)
	for _, v := range tokens {
		if v == "+" {
			stack = append(stack[:len(stack)-2], stack[len(stack)-2]+stack[len(stack)-1])
		} else if v == "-" {
			stack = append(stack[:len(stack)-2], stack[len(stack)-2]-stack[len(stack)-1])
		} else if v == "*" {
			stack = append(stack[:len(stack)-2], stack[len(stack)-2]*stack[len(stack)-1])
		} else if v == "/" {
			stack = append(stack[:len(stack)-2], stack[len(stack)-2]/stack[len(stack)-1])
		} else {
			t, _ := strconv.Atoi(v)
			stack = append(stack, t)
		}
	}
	return stack[0]
}

File: cpp/0150M-逆波兰表达式求值.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-18 09:35:56
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    int evalRPN(vector<string>& tokens) {
        stack<int> st;
        int s1, s2;
        for (auto s : tokens) {
            if (s == "+" || s == "-" || s == "*" || s == "/") {
                s1 = st.top();
                st.pop();
                s2 = st.top();
                st.pop();
                if (s == "+")
                    st.push(s2 + s1);
                else if (s == "-")
                    st.push(s2 - s1);
                else if (s == "*")
                    st.push(s2 * s1);
                else if (s == "/")
                    st.push(s2 / s1);
            } else
                st.push(stoi(s));
        }
        return st.top();
    }
};